File: lua/day3_part1_2020.lua

local file = io.open("input.txt", "r")
local forest = {}
for line in file:lines() do
    table.insert(forest, line)
end
file:close()

function countTrees(forest, right, down)
    local trees = 0
    local x = 0
    local width = #forest[1]

    for y = 1, #forest, down do
        if string.sub(forest[y], x % width + 1, x % width + 1) == "#" then
            trees = trees + 1
        end
        x = x + right
    end

    return trees
end

local trees = countTrees(forest, 3, 1)
print(trees)

File: clojure/day3_part1_2020.clj
(import '[java.io BufferedReader FileReader])

(defn read-lines [filename]
  (with-open [reader (BufferedReader. (FileReader. filename))]
    (loop [lines []]
      (let [line (.readLine reader)]
        (if line
          (recur (conj lines line))
          lines)))))

(defn count-trees [forest right down]
  (let [width (count (first forest))
        height (count forest)]
    (loop [trees 0
           x 0
           y 0]
      (if (< y height)
        (let [current-row (nth forest y)
              current-cell (get current-row (mod x width))]
          (recur (+ trees (if (= current-cell \#) 1 0))
                 (+ x right)
                 (+ y down)))
        trees))))

(def forest (read-lines "input.txt"))
(def trees (count-trees forest 3 1))
(println trees)